### PR TITLE
Throw Hopcroft-Karp at shapeless craft recipes

### LINKED ISSUE #11224: Shapeless crafts causing lag
##### Minetest version
Minetest 5.4

##### OS / Hardware
Lubuntu 64-bit 21.04 using AMD CPU/GPU

##### Summary
It seems that adding to many 'shapeless' crafts can cause lag on a server, at one point while crafting we had 36 seconds of lag on Xanadu and after removing all shapeless crafts from my mods it disappeared.

##### Steps to reproduce
Add mods with loads of shapeless crafts and try crafting a chest with different types of wood.

##### Maybe related
https://github.com/minetest/minetest/pull/8097

### PR DESCRIPTION
**Pull request by Desour**
Fixes performance for shapeless recipes with groups.
Fixes #11224.

The reason why it was so slow is that the craft check tries out all permutations of the recipe. In the worst case, that's 9! = 362880 possibilities.
In the best case, there are no groups and the first permutation matches because recipe and input are sorted. And the prefiltering works perfectly, so the check will always pass.
If there is at least one group however, the prefiltering just filters by count, and well, it gets bad. In the test code, I've used many different items, to create more permutations.

Fixing this is not hard. Each input item can satisfy some recipe slots. If you imagine the input and recipe slots as nodes in a graph, and the satisfaction relationships as edges, you get a bipartite graph. We just have to find out if we can find a matching in this bipartite graph that matches every node. Finding a maximum-cardinality matching in a bipartite graph is easy, i.e. by using the hopcroft-karp algorithm. See <https://en.wikipedia.org/w/index.php?title=Hopcroft%E2%80%93Karp_algorithm&useskin=vector>.

I've also added code to filter out non-recipe slots, as that's simply a cheap sorted merge.

Note that there can still be done more to do for faster crafting (it's still super slow, i.e. 200us for a single craft check), i.e. better filtering, and also caching. I've already planned to do a PR for the former.

## To do

This PR is a Ready for Review.

For reviewing, you can compare my code with the pseudocode on wikipedia, as it's almost the same: <https://en.wikipedia.org/w/index.php?title=Hopcroft%E2%80%93Karp_algorithm&useskin=vector#Pseudocode>

### For performance:
```lua
minetest.after(1, function()
	minetest.log("[bench_craft] doing...")

	local input = {
			method = "normal",
			width = 3,
			items = {
				"default:stone", "default:cobble", "default:cobble",
				"default:stone", "default:cobble", "default:cobble",
				"default:stone", "default:cobble", "default:cobble",
			},
		}

	local t0 = minetest.get_us_time() * 1e-6

	minetest.get_craft_result(input)

	local t1 = minetest.get_us_time() * 1e-6

	minetest.log("[bench_craft] it took "..((t1-t0) * 1e6).." us")
end)

-- comment this out to compare the difference
minetest.register_craft({
	type = "shapeless",
	output = "default:dirt",
	recipe = {
		"group:wood", "default:tree", "default:stick",
		"default:leaves", "default:dirt", "default:dirt_with_grass",
		"group:stone", "default:cobble", "default:cactus",
	},
})
```
In master, the recipe adds about 80 ms (milliseconds!) to the single `get_craft_result` call.
With the PR, the added recipe makes no significant difference.

### For correctness:
* Look through the code.
* Try out your favourite craft recipes.
* Run the unittest: `$ minetest --run-unittests`

I've also compared the results of my implementation of the algorithm with a the results of a lib that can compute the same: <https://luarocks.org/modules/chen0040/luagraphs>
(I first wanted to compare to boost, but that's too complicated for me now.)
To run the test, follow the following steps:
* Install luagraphs: luarocks install luagraphs.
* Fetch and build this branch: <https://github.com/Desour/minetest/tree/craft_bip_match_test_compare>
* Create a new mod with the following code in init.lua:
<details>

```lua

local ffi = require("ffi")

ffi.cdef([[
int ffi_hopcroft_karp_num_matches(const bool *bip_graph_arr, int graph_size);
]])

local luagraphs_data_network = require('luagraphs.data.network')
local luagraphs_flow_FordFulkerson = require('luagraphs.flow.FordFulkerson')

local function max_flow_my(matrix, size)
	local bip_graph_arr = ffi.new("bool[81]", matrix)

	return ffi.C.ffi_hopcroft_karp_num_matches(bip_graph_arr, size)
end

local function max_flow_luagraphs(matrix, size)
	local num_nodes = 1+size+size+1
	local s = 0
	local t = num_nodes-1
	local u_min = 1
	local u_max = u_min+size-1
	local v_min = u_max+1
	local v_max = v_min+size-1
	assert(v_max+1 == t)

	local g = luagraphs_data_network.FlowNetwork.create(num_nodes)

	for u = u_min, u_max do
		g:addEdge(s, u, 1)
	end
	for v = v_min, v_max do
		g:addEdge(v, t, 1)
	end

	for u = 0, size-1 do
		for v = 0, size-1 do
			if matrix[u*9+v] then
				g:addEdge(u_min+u, v_min+v, 1)
			end
		end
	end

	local method = luagraphs_flow_FordFulkerson.create()
	return method:run(g, s, t)
end


minetest.after(1, function()
	minetest.log("sarting...")
	math.randomseed(35134) -- change this

	local matrix = {}
	for i = 0, 81-1 do
		matrix[i] = false
	end

	for i = 1, 10000 do
		local size = math.random(9)
		local fill_rate = math.random()

		for u = 0, size-1 do
			for v = 0, size-1 do
				matrix[u*9+v] = math.random() > fill_rate
			end
		end

		local result_my = max_flow_my(matrix, size)
		local result_luagraphs = max_flow_luagraphs(matrix, size)
		if result_my ~= result_luagraphs then
			minetest.log("oh no!")
			minetest.log(string.format("%s ~= %s", result_my, result_luagraphs))
			--~ minetest.log(dump({size = size, matrix = matrix}))
			return
		end
	end

	minetest.log("all fine.")
end)

```

</details>

* Optional: Change the randomseed.
* Enable the mod.
* Disable mod security.
* Start the world.
* Wait a second and look into chat. Expected: "all fine."
* Don't forget to enable mod security again.

### CODE REVIEW

**Comment by fluxionary** [comment]
i've tested the sample code from #13231, and get the much-more-reasonable result of "get_craft_result: 9 wheat -> 1 mulch took .0006s". bravo :)

**Comment by Desour** [comment]
I've compared the `hopcroft_karp_can_match_all` to the results of this library (by creating random graphs): <https://github.com/chen0040/lua-graph> (See top comment for details.)
I found no flaws.

**Comment by lhofhansl** [comment]
Is this waiting from something?

**Comment by Desour** [comment]
No, this is finished.
(More testing would probably not hurt though.)

**Comment by fluxionary** [comment]
> Is this waiting from something?

i certainly don't want to hold this up, but is there some reason it's still labeled as helping shapeless crafting recipes? it seems to help shaped crafting recipes just as well.

**Comment by Desour** [comment]
>is there some reason it's still labeled as helping shapeless crafting recipes? it seems to help shaped crafting recipes just as well.

The code for shaped craft checking isn't touched at all. If you only register shaped recipes, there should be no difference. Note that input grids are can be both shaped and shapeless, so minetest always checks all sorts of recipes (in some cases at least (lower priority recipes can be ignored if they come later)).

**Comment by SmallJoker** [comment]
Testing mod N°133
```Lua
minetest.register_on_joinplayer(function(player)
	local inv = player:get_inventory()
	inv:set_size("craft", 4*4)
	inv:set_width("craft", 4)
end)

minetest.register_craft({
    type = "shapeless",
    output = 'default:mese',
    recipe = {
		-- 5x stick, 4x apple, 3x stone, 1x water bucket
        "default:stick",
        "default:stick",
        "default:apple",
        "default:stick",
        "default:apple",
        "bucket:bucket_water",
        "default:stick",
        "default:apple",
        "default:stone",
        "default:stick",
        "default:apple",
        "default:stone",
        "default:stone",
    },
	replacements = {
		{ "bucket:bucket_water", "bucket:bucket_empty" },
		{ "default:stone", "default:cobble 4" },
	}
})

-- unified_inventory patch:
	style.craft_grid =	table.concat({
							ui.make_inv_img_grid(style.craft_x, style.craft_y, 4, 4),
							ui.single_slot(style.craft_x + ui.imgscale*4, style.craft_y), -- the craft result slot
							string.format("image[%f,%f;%f,%f;ui_crafting_arrow.png]",
							style.craft_arrow_x, style.craft_y, ui.imgscale, ui.imgscale),
							string.format("list[current_player;craft;%f,%f;4,4;]",
```

Works.

**Review by SmallJoker — APPROVED** [review]
Tested shapeless 4x4 recipe. No issues found.

**Comment by Desour** [comment]
Recipes with >9 groups are not supported. Marked as draft until fixed.

File: src/craftdef.cpp [hunk 1]
```diff
@@ -493,6 +493,113 @@ std::string CraftDefinitionShapeless::getName() const
...
+// See https://en.wikipedia.org/w/index.php?title=Hopcroft-Karp_algorithm for
+// details.
+static bool hopcroft_karp_can_match_all(const std::vector<std::vector<u16>> &bip_graph)
+{
+	assert(bip_graph.size() <= SHAPELESS_GROUPS_MAX);
+	u16 graph_size = bip_graph.size();
+	const u16 nil = graph_size; // nil / dummy index
+	constexpr u16 inf = UINT16_MAX; // bigger than any path length (> SHAPELESS_GROUPS_MAX * 2)
+
+	auto pair_u = std::make_unique<u16[]>(graph_size + 1); // for each u (or nil) the matched v (or nil)
```
Comment (SmallJoker) [inline src/craftdef.cpp:510]: I suppose you could've used vectors + reserve for proper memory management. Well, this works too.

Comment (Desour) [reply src/craftdef.cpp:510]: I first had it at vectors + resize. But it only needs fixed-sized arrays, so I thought this was prettier. `¯\_(ツ)_/¯`

Thanks for review, btw.! ^^

**Comment by Desour** [comment]
Modified the unittests a tiny little to also check aliases. Should not affect approval.

**Comment by Desour** [comment]
(Manually squashed into two commits before merge.)

**Pull request merged**